### PR TITLE
feat: refactor manifest to get snapshot in memory

### PR DESCRIPTION
**Pull request by Rachelint**
## Which issue does this PR close?

Closes #
Part of #799 

## Rationale for this change
 Make snapshot able to get snapshot from table data.

## What changes are included in this PR?
Make snapshot able to get snapshot from table data.

## Are there any user-facing changes?
None.

## How does this change test
Test by ut.

### CODE REVIEW

File: analytic_engine/src/manifest/details.rs [hunk 1]
```diff
@@ -317,12 +524,14 @@ impl Manifest for ManifestImpl {
...
-            location,
+        let storage_snapshot_builder = StorageSnapshotBuilder {
             log_store,
             snapshot_store,
         };
-        let snapshot = snapshotter.create_latest_snapshot().await?;
+        let snapshotter = Snapshotter {
+            snapshot_builder: storage_snapshot_builder,
+        };
+        let snapshot = snapshotter.snapshot().await?;
```
Comment (ShiKaiWi) [inline analytic_engine/src/manifest/details.rs:534]: In the previous implementation, only the latest snapshot is fetched but won't be stored and the old logs would also be kept. If so, a snapshot will be generated and stored every time read the table manifest data, and I guess it is unacceptable.

**Review by ShiKaiWi — COMMENTED** [review]
This solution sounds good. However, there may be one potential problem in it. Here is an example for it:
1. The manifest wal is updated;
2. Snapshot is triggered;
3. Table metadata in the memory is used to form the latest snapshot;

In the step 3, the table metadata (snapshot) may be different from the logs from manifest wal.

File: analytic_engine/src/manifest/details.rs [hunk 2]
```diff
@@ -588,99 +791,15 @@ impl From<Snapshot> for manifest_pb::Snapshot {
...
-            Ok(Some(Snapshot {
-                end_seq: latest_seq,
-                original_logs_num: num_logs,
-                data: manifest_data_builder.build(),
-            }))
-        } else {
-            Ok(None)
-        }
+    pub async fn snapshot(&self) -> Result<Option<Snapshot>> {
+        self.snapshot_builder.build().await
```
[the remark this answers is not in the document]
Comment (Rachelint) [reply analytic_engine/src/manifest/details.rs]: I have splitter `Snapshotter` to `Snapshotter` and `SnapshotterRecoverer`, and removed the `SnapshotBuilder` now.

**Comment by Rachelint** [comment]
> This solution sounds good. However, there may be one potential problem in it. Here is an example for it:
> 
>     1. The manifest wal is updated;
> 
>     2. Snapshot is triggered;
> 
>     3. Table metadata in the memory is used to form the latest snapshot;
> 
> 
> In the step 3, the table metadata (snapshot) may be different from the logs from manifest wal.

In fact, this situation will not occurred.
Snapshot works always done before writing the new comming logs to log store.

File: analytic_engine/src/manifest/details.rs [hunk 3]
```diff
@@ -155,6 +162,192 @@ impl MetaUpdateLogEntryIterator for MetaUpdateReaderImpl {
...
+        &self,
+        space_id: SpaceId,
+        table_id: TableId,
+    ) -> Result<Option<TableManifestData>>;
+}
+
+#[derive(Clone)]
+pub(crate) struct TableSnapshotProviderImpl {
+    pub(crate) spaces: SpacesRef,
+}
```
Comment (ShiKaiWi) [inline analytic_engine/src/manifest/details.rs:179]: Maybe we should put this implementation to the space or table module?

**Comment by ShiKaiWi** [comment]
> > This solution sounds good. However, there may be one potential problem in it. Here is an example for it:
> > ```
> > 1. The manifest wal is updated;
> > 
> > 2. Snapshot is triggered;
> > 
> > 3. Table metadata in the memory is used to form the latest snapshot;
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > In the step 3, the table metadata (snapshot) may be different from the logs from manifest wal.
> 
> In fact, this situation will not be occurred. Snapshot works always done before writing the new comming logs to log store.

After some thoughts, I'm still worried about that the table metadata in the memory is not controlled by the manifest, that is to say, the order of updating manifest before the table metadata in memory is implicit. Maybe we can extend the TableSnapshotProvider to TabeSnapshotManager which is not only able to provide the latest snapshot but also receive metadata updates.

**Comment by Rachelint** [comment]
> > > This solution sounds good. However, there may be one potential problem in it. Here is an example for it:
> > > ```
> > > 1. The manifest wal is updated;
> > > 
> > > 2. Snapshot is triggered;
> > > 
> > > 3. Table metadata in the memory is used to form the latest snapshot;
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > In the step 3, the table metadata (snapshot) may be different from the logs from manifest wal.
> > 
> > 
> > In fact, this situation will not be occurred. Snapshot works always done before writing the new comming logs to log store.
> 
> After some thoughts, I'm still worried about that the table metadata in the memory is not controlled by the manifest, that is to say, the order of updating manifest before the table metadata in memory is implicit. Maybe we can extend the TableSnapshotProvider to TabeSnapshotManager which is not only able to provide the latest snapshot but also receive metadata updates.

I will sovle it in next pr.

**Review by ShiKaiWi — APPROVED** [review]
LGTM

**Pull request merged**